docs/design/proposals/rbd-volume-healer.md

```diff
@@ -0,0 +1,102 @@
+at Ceph-CSI we will also have a userspace way of mounting the rbd images,
+via rbd-nbd.
+
+[Rbd-nbd](https://docs.ceph.com/en/latest/man/8/rbd-nbd/) is a client for
+RADOS block device (rbd) images like the existing rbd kernel module. It
+will map an rbd image to an nbd (Network Block Device) device, allowing
+access to it as a regular local block device.
+
+![csi-rbd-nbd](./images/csi-rbd-nbd.svg)
+
+It’s worth making a note that the rbd-nbd processes will run on the
+client-side, which is inside the `csi-rbdplugin` node plugin.
+
+### Advantages of userspace mounters
+
+- It is easier to add features to rbd-nbd as it is released regularly with
+  Ceph, and more difficult and time consuming to add features to the kernel
+  rbd module as that is part of the Linux kernel release schedule.
+- Container upgrades will be independent of the host node, which means if
+  there are any new features with rbd-nbd, we don’t have to reboot the node
+  as the changes will be shipped inside the container.
+- Because the container upgrades are host node independent, we will be a
+  better citizen in K8s by switching to the userspace model.
+- Unlike krbd, rbd-nbd uses librbd user-space library that gets most of the
+  development focus, and hence rbd-nbd will be feature-rich.
+- Unlike rbd-nbd, krbd exhibits additional drawback, being entirely kernel
+  space impacts fault-tolerance as any kernel panic affects a whole node
+  not only a single pod that is using rbd storage. Thanks to the rbd-nbd’s
+  userspace design, this doesn’t bother here. One might also argue that
+  rbd-nbd has to work with NBD module closely, yes! but remember it's
+  mostly unchanged much from years and consider it to be much stable now.
+  NBD is a generic driver and there will be many eyes on it compared to the
+  rbd driver. rbd driver is vendor-specific logic that might also need
+  changes every now and then when a new feature update is available at Ceph.
+
+### Side effects of userspace mounters
+
+Since the rbd-nbd processes run per volume map on the client side i.e.
+inside the `csi-rbdplugin` node plugin, a restart of the node plugin will
+terminate all the rbd-nbd processes, and there is no way to restore
+these processes back to life currently, which could lead to IO errors
+on all the application pods.
+
+![csi-plugin-restart](./images/csi-plugin-restart.svg)
+
+This is where the Volume healer could help.
+
+## Volume healer
+
+Volume healer runs on the start of rbd node plugin and runs within the
+node plugin driver context.
+
+Volume healer does the below,
+
+- Get the Volume attachment list for the current node where it is running
+- Filter the volume attachments list through matching driver name and
+  status attached
+- For each volume attachment get the respective PV information and check
+  the criteria of PV Bound, mounter type
+- Build the StagingPath where rbd images PVC is mounted, based on the
+  KUBELET path and PV object
```

Comment:
This leaks Kubernetes implementation details to CSI drivers. CSI drivers should be able to run outside of Kubernetes.
Can you store a state of the driver on the host (say in `/var/lib/rbd-csi/*`, available in the driver pod as HostPath volume)? With a file for each rbd-nbd to restore after restart (or whatever you need, but let it be somewhat transactional to survive pod/machine restarts safely).

Reply:
Given the changes are already merged and it's an alpha release, I will open an issue for this to make further improvements and target any better suggestions for the beta.

Reply:
Placing the required details in our own host-path volume is not so much an issue, we can change that.

However, we would also require the credentials to access the Ceph cluster. In addition to security concerns, it is possible that these expire and are not updated (would get updated in the Kubernetes Secrets from the StorageClass). So at least some part of the healer is dependent on Kubernetes at the moment (calling the NodeStageVolume). We're all not happy about the Kubernetes dependency, but for a few other features we already have that as well :disappointed: